repository/src/lib.rs

```diff
@@ -196,8 +196,88 @@ impl<T: RawRepository> DistributedRepository<T> {
-    pub async fn sync(&mut self, _block_commit: &CommitHash) -> Result<(), Error> {
-        unimplemented!()
+    pub async fn sync(&mut self, block_commit: &CommitHash) -> Result<(), Error> {
+        // See if this commit is a descendant of the current finalized branch
+        let current_finalized_commit = self.raw.locate_branch(FINALIZED_BRANCH_NAME.into()).await?;
+        if self
+            .raw
+            .find_merge_base(current_finalized_commit, *block_commit)
+            .await?
+            != current_finalized_commit
+        {
+            return Err(anyhow!(
+                "block commit is not a descendant of the current finalized branch"
+            ));
+        }
+
+        // Construct a commit list starting from the next commit of the last finalized block to the `branch_tip_commit`(the most recent commit of the branch)
+        let ancestor_commits = self.raw.list_ancestors(*block_commit, Some(256)).await?;
+        let position = ancestor_commits
+            .iter()
+            .position(|c| *c == current_finalized_commit)
+            .expect("TODO: handle the case where it exceeds the limit");
+        let new_semantic_commits = stream::iter(
+            ancestor_commits
+                .iter()
+                .take(position)
+                .rev()
+                .cloned()
+                .map(|c| {
+                    let raw = &self.raw;
+                    async move { raw.read_semantic_commit(c).await.map(|x| (x, c)) }
+                }),
+        )
+        .buffered(256)
+        .collect::<Vec<_>>()
+        .await;
+        let mut new_semantic_commits = new_semantic_commits
+            .into_iter()
+            .collect::<Result<Vec<_>, _>>()?;
+        // Add most recent commit of the branch to the list since it is not included in the ancestor commits
+        let block_semantic_commit = self.raw.read_semantic_commit(*block_commit).await?;
+        new_semantic_commits.push((block_semantic_commit, *block_commit));
+        let new_commits = new_semantic_commits
+            .into_iter()
+            .map(|(commit, hash)| {
+                from_semantic_commit(commit)
+                    .map_err(|e| (e, hash))
+                    .map(|x| (x, hash))
+            })
+            .collect::<Result<Vec<_>, _>>()
+            .map_err(|(error, hash)| anyhow!("failed to convert the commit {}: {}", hash, error))?;
+
+        let start_header = self.get_last_finalized_block_header().await?;
+        let reserved_state = self.get_reserved_state().await?;
+
+        // Verify all the new commits
+        let mut verifier =
+            CommitSequenceVerifier::new(start_header.clone(), reserved_state.clone())
+                .map_err(|e| anyhow!("failed to create a verifier: {}", e))?;
+        for (new_commit, _) in &new_commits {
+            if verifier.apply_commit(new_commit).is_err() {
+                return Err(anyhow!("CSV failed"));
+            }
+        }
```

Comment:
```suggestion
        for (new_commit, new_commit_hash) in &new_commits {
            verifier
                .apply_commit(new_commit)
                .map_err(|e| anyhow!("verification error on commit {}: {}", new_commit_hash, e)?;
        }
```
please keep the implementation consistent

Reply:
fixed.